### PR TITLE
Dense Layer, Sigmoid Operation, Overloads and Updated Tensor Operation Calls

### PR DESCRIPTION
**Pull request by karanchahal**
This is a big pull request ! So let's strap in 

1. Added the functionality of the dense layer/ fully connected layer. Right now, weights are being initialised with the Glorot Initialisation and the default activation function is Sigmoid.

2. The functionality of the dot product has been added . CPU only in this pull request. We support the back propagation through the dot product too now. 

3. Added the standalone sigmoid Operation with it's supported backPropagation. There is a small numeric difference between this sigmoid operation and the manually  calculated one. I will file a bug for this to mark for further debugging. 

4. Separated all matrix, tensor and vector operations to seperate folder called overloads/

5. Updated recommended way to perform operations between Tensors is through tensorOps::add(x,y) etc. Support for all operations is in tensorOps. it operates using pointers so it preserves the computational graph. This is the recommended way going forward.

6. Wrote tests for all functionality mentioned above.  

7. Configured CI/CD with Circle CI. Now tests will be run on each commit of a pull request !

### CODE REVIEW

**Comment by karanchahal** [comment]
@uditarora As the pull request is really big. Cloning this branch and going through the code will be beneficial :) Let's get this merged by today so that we can work on SGD and loss functions !

File: layers/dense.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        For input x, it gets output y.
+        Where, y = activation(weights@x + biases)
+
+        Where @ represents dot product.
+        other operations are all elementwise.
+    */
+    Tensor<T>* forward(Tensor<T> *x) {
+
+
+        auto dot = tensorOps::dot(x,weights);
```
Comment (uditarora) [inline layers/dense.h:85]: nit: whitespace between args

File: layers/dense.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        auto z = tensorOps::add(dot, biases);
+        Tensor<T>* logits;
+        
+        if(act == SIGMOID) {
+            logits = tensorOps::sigmoid(z);
+        }
```
Comment (uditarora) [inline layers/dense.h:93]: Should we also handle the case where we don't want to add any activation?

Perhaps by adding an else here, and assigning logits equal to z?

File: operations/addOperation_Impl.h [hunk 3]
```diff
@@ -1,24 +1,45 @@
...
 #define __OP_ADD_IMPL_INCLUDED__  
 
+/* 
+    Backpropogation of the addition operation. The addition operation distributes the gradient. So it
+    effectively just transfers the gradient coming in to the various inpuot sources.
+*/
 template <typename T>
 void AddOperation<T>::backward(Matrix<T> grad) {
-    // Distributing case: where gradients are backproped as is
+
```
Comment (uditarora) [inline operations/addOperation_Impl.h:17]: nit: remove newline

File: operations/addOperation_Impl.h [hunk 4]
```diff
@@ -1,24 +1,45 @@
...
 template <typename T>
 Tensor<T> AddOperation<T>::forward() {
     this->t3 = new Tensor<T>(this->t1->val + this->t2->val, this);
     return *this->t3;
 }
 
+/* 
+    Forward Propogation of the operation. Return pointer to the tensor.
+*/
+template <typename T>
```
Comment (uditarora) [inline operations/addOperation_Impl.h:39]: Going forward, let's make the pointer based op the default, thereby just calling it "forward", and the other one as forwardDeprecated or something?

File: operations/addOperation.h [hunk 5]
```diff
@@ -11,10 +11,8 @@ two tensors.
 template <typename T>
 class AddOperation : public Operation<T> {
     public:
-    Tensor<T> *t1 = NULL, *t2 = NULL; // Input tensors
-    Tensor<T> *t3 = NULL; // Output tensor
-
-    AddOperation(Tensor<T> *t1, Tensor<T> *t2) : Operation<T>(t1,t2){
+    
+    AddOperation(Tensor<T> *t1, Tensor<T> *t2) {
```
Comment (uditarora) [inline operations/addOperation.h:15]: For my understanding - why has the call to the base class init function been removed here?

Comment (uditarora) [reply operations/addOperation.h:15]: And from other operations

File: operations/exponentOperation.h [hunk 6]
```diff
@@ -1,6 +1,6 @@
 /*
-This file defines the ExponentOperation class which represents the
-exponentiation of a tensor.
+    This file defines the ExponentOperation class which represents the
+    exponentiation of a tensor.
```
Comment (uditarora) [inline operations/exponentOperation.h:3]: General comment about this operation - We will end up allocating pointer space for one more tensor pointer here (t2) because it extends from "Operation" which declares 3 pointers in memory.

One possible solution to avoiding this can be to have two different types of operations extending from the base "Operation" class - "BinaryOperation" and "UnaryOperation" and then having our operations extending from either of these. I can take that up as a separate issue if you want.

File: operations/sigmoidOperation_Impl.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    
+    F = sigmoid(x) is forward propogation
+    The gradients would be as follows:
+    1. dF/dx = (1 - F)*F
+
+    This formula can be derived by hand.
+*/
+template <typename T>
+void SigmoidOperation<T>::backward(Matrix<T> grad) {
+    auto g = ((T)1 + ((T)-1)*this->t3->val)*this->t3->val; // matrix can be overloaded
```
Comment (uditarora) [inline operations/sigmoidOperation_Impl.h]: Can we make this line of code simpler so that it is easier to understand?

Maybe by splitting it into multiple lines.

File: overloads/matrix.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+// Overloaded function for printing matrix: cout<<matrix<<endl;
+template<typename T>
+ostream & operator << (ostream &out, Matrix<T> &m) {
+    vector<int> stack;
+    return m.print(out,stack,0);
+}
+
+// Divison with a scalar as divident
+template<typename E>
```
Comment (uditarora) [inline overloads/matrix.h]: nit: Any particular reason not for using 'T' in the typename arg here?

File: overloads/tensor.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        one->frontOp = new MultiplyOperation<T>(one, two);
+        two->frontOp = one->frontOp;
+        return one->frontOp->forwardPointer();
+    }
+
+    // Multiply with scalar
+    template<typename T>
+    Tensor<T>* multiply(T v, Tensor<T>* two) {
+        auto one = new Tensor<T>(vector<T>(two->val.val.size(),v),two->val.shape);
+        return multiply(one,two);
```
Comment (uditarora) [inline overloads/tensor.h:54]: nit: whitespace between args

File: overloads/tensor.h [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        one->frontOp = new AddOperation<T>(one, two);
+        two->frontOp = one->frontOp;
+        return one->frontOp->forwardPointer();
+    }
+
+    // Addition with Scalar
+    template<typename T>
+    Tensor<T>* add(T v, Tensor<T>* two) {
+        auto one = new Tensor<T>(vector<T>(two->val.val.size(),v),two->val.shape);
+        return add(one,two);
```
Comment (uditarora) [inline overloads/tensor.h:24]: nit: whitespace between args

File: overloads/tensor.h [hunk 11]
```diff
@@ -0,0 +1,81 @@
+/*
+    This file defines the various mathematical overloads of the Tensor Class.
```
Comment (uditarora) [inline overloads/tensor.h]: This class doesn't define "overloads" as the comment mentions. It defines "operations".

File: overloads/vector.h [hunk 12]
```diff
@@ -74,5 +108,24 @@ vector<T> exponent(const vector<T> &a) {
     return arr;
 }
 
+// isEquals operator
+template<typename T>
+bool operator == (const vector<T> &a, const vector<T> &b) {
+    int n = a.size();
```
Comment (uditarora) [inline overloads/vector.h]: supernit: vector.size() returns an unsigned int, but we're typecasting it to int here. If there is a scenario where this typecasting leads to overflow, this condition might fail even though the vectors have the same size. A cleaner way to check would be to just have the size method calls inside the if condition directly.

File: types/tensor.h [hunk 13]
```diff
@@ -1,7 +1,21 @@
 /*
-This file defines the Tensor class which stores the matrices used for
-computations by the library, as well as the forward and backward operations
-being executed on it in the form of pointers to Operation objects.
+    This file defines the Tensor class. The Tensor is the base object used for creating
```
Comment (uditarora) [inline types/tensor.h:2]: Love the detailed comments :)

File: types/tensor.h [hunk 14]
```diff
@@ -10,31 +24,61 @@ being executed on it in the form of pointers to Operation objects.
...
+    /*
+        This function is called during the initilaisation of Tensor. It sets the value of it's gradients to zero. This is needed as 
+        during backPropogation the same tensor can be used for different operation, hence to calculate it's partial gradients
+        each individual operation's gradients have to be summed up. Hence we initialise the tensor's gradients to zero.
+        
+        See constructor for it's usage.
+    */
+    void zeroGrad() {
+        assert(val.shape.size() != 0 && "The value of matrix cannot be uninitialised during initialisng zeros in tensor's gradient");
+        vector<T> g(val.val.size(),0);
```
Comment (uditarora) [inline types/tensor.h]: nit: whitespace between args

File: operations/addOperation.h (hunk 5 above)
Comment (karanchahal) [reply operations/addOperation.h:15]: This is  because I've removed the t1 ,t2 and t3 declarations from the inherited class. now whenever I refer to t1,t2 and t3. They refer to the variables in the base Operation class. Hence, there is no need for the Operation(t1,t2) call of the base class to set them as they are already set.

File: overloads/vector.h (hunk 12 above)
Comment (karanchahal) [reply overloads/vector.h]: Just converted the n and m to type size_t. Now the size will not overflow as size_t is the base numeric data type used throughout C++.  It is the base class of all float, double, int values.

File: types/tensor.h (hunk 13 above)
Comment (karanchahal) [reply types/tensor.h:2]: :)

File: operations/addOperation_Impl.h (hunk 4 above)
Comment (karanchahal) [reply operations/addOperation_Impl.h:39]: Done

File: overloads/tensor.h (hunk 10 above)
Comment (karanchahal) [reply overloads/tensor.h:24]: Done

File: operations/addOperation_Impl.h (hunk 3 above)
Comment (karanchahal) [reply operations/addOperation_Impl.h:17]: Done

File: layers/dense.h (hunk 2 above)
Comment (karanchahal) [reply layers/dense.h:93]: Done, I added a new enum called NO_ACTIVATION

File: layers/dense.h (hunk 1 above)
Comment (karanchahal) [reply layers/dense.h:85]: Done

File: operations/exponentOperation.h (hunk 6 above)
Comment (karanchahal) [reply operations/exponentOperation.h:3]: Yeah, we could do this but I think it'll add unnecessary complexity to the code. I think we should keep it the way it is. I'm all for memory saving but also want to keep the code to as few lines as possible. One tensor pointer takes really minimal space so I think we should be good.

File: operations/sigmoidOperation_Impl.h (hunk 7 above)
Comment (karanchahal) [reply operations/sigmoidOperation_Impl.h]: I think it's pretty clear. And we're doing operator overloading to make expressions like these to fit in one line. I've added a comment that signifies what expression it is to make it clearer.

File: operations/exponentOperation.h (hunk 6 above)
Comment (uditarora) [reply operations/exponentOperation.h:3]: Alright, that's fair.

**Review by uditarora — APPROVED** [review]

File: overloads/matrix.h (hunk 8 above)
Comment (karanchahal) [reply overloads/matrix.h]: Yes, fixed. Changed to T everywhere.

File: overloads/tensor.h (hunk 9 above)
Comment (karanchahal) [reply overloads/tensor.h:54]: Done

File: overloads/tensor.h (hunk 11 above)
Comment (karanchahal) [reply overloads/tensor.h]: Done

**Comment by karanchahal** [comment]
Looks good now ! Merging :+1:

**Comment by karanchahal** [comment]
Thanks @uditarora  for the great review !
